### PR TITLE
fix: [CI-5347]: Add support for custom handling of Bitbucket Server API URL

### PR DESCRIPTION
**Pull request by rutvijmehta-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/36695)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/36695)

File: 960-api-services/src/main/java/io/harness/git/GitClientHelper.java [hunk 1]
```diff
@@ -237,10 +237,37 @@ public static String getBitBucketApiURL(String url) {
       return "https://api.bitbucket.org/";
     } else {
       String domain = GitClientHelper.getGitSCM(url);
+      domain = fetchCustomBitbucketDomain(url, domain);
       return getHttpProtocolPrefix(url) + domain + "/";
     }
   }
 
+  private static String fetchCustomBitbucketDomain(String url, String domain) {
+    String SCM_SPLITTER = "/scm";
```
Comment (harshjain12) [inline 960-api-services/src/main/java/io/harness/git/GitClientHelper.java:246]: make it final

File: 960-api-services/src/test/java/io/harness/git/GitClientHelperTest.java [hunk 2]
```diff
@@ -503,6 +503,22 @@ public void testGetBitBucketApiURL() {
         .isEqualTo("https://api.bitbucket.org/");
     assertThat(GitClientHelper.getBitBucketApiURL("http://10.67.0.1/devkimittal/harness-core"))
         .isEqualTo("http://10.67.0.1/");
+    assertThat(GitClientHelper.getBitBucketApiURL("https://bb.harness.io/stash/scm/rutvijproject/rutvijrepo.git"))
```
Comment (harshjain12) [inline 960-api-services/src/test/java/io/harness/git/GitClientHelperTest.java:506]: Add test cases for ssh url

File: 960-api-services/src/main/java/io/harness/git/GitClientHelper.java (hunk 1 above)
Comment (rutvijmehta-harness) [reply 960-api-services/src/main/java/io/harness/git/GitClientHelper.java]: Done

File: 960-api-services/src/test/java/io/harness/git/GitClientHelperTest.java (hunk 2 above)
Comment (rutvijmehta-harness) [reply 960-api-services/src/test/java/io/harness/git/GitClientHelperTest.java:506]: Done

**Pull request merged**